### PR TITLE
fix(core): select component reading the selected option and correct tab order to access the element in mobile mode

### PR DESCRIPTION
**Pull request by manu-kr**
#### Please provide a link to the associated issue.
Closes https://github.com/SAP/fundamental-ngx/issues/5051

#### Please provide a brief summary of this pull request.
changes to fix for Select accessibility bug reading the selected option and also fix the correct tab order to access the element in mobile mode

#### Please check whether the PR fulfils the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [x] Run npm run build-pack-library and test in external application

Documentation checklist:
- [NA] update `README.md`
- [NA] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Review by Betrozov — APPROVED** [review]

**Comment by sKudum** [comment]
It reads out the name but does not mention that item is selected on select.


when I select apple
onselect reads out now=>  apple

expected message=> apple is selected or selected apple.

We can see examples of angular material examples for the messages.

**Review by sKudum — CHANGES_REQUESTED** [review]
1)It reads out the name but does not mention that item is selected on select.

when I select apple
onselect reads out now=> apple

expected message=> apple is selected or selected apple.

2).In mobile mode, on tab we can see 2 focus. it has to be 1 at a time.

**Comment by manu-kr** [comment]
> 1)It reads out the name but does not mention that item is selected on select.
> 
> when I select apple
> onselect reads out now=> apple
> 
> expected message=> apple is selected or selected apple.
> 
> 2).In mobile mode, on tab we can see 2 focus. it has to be 1 at a time.
> 
> 3).On tab we can see focus with blue border on message. Is it expected ?
> 

> Now its reading properly.

>In Mobile mode Tab key won't present so we are discarding tab key access for mobile after. discussing with Sanjay , Puru and Sushma.

**Comment by sKudum** [comment]
> > 1)It reads out the name but does not mention that item is selected on select.
> > when I select apple
> > onselect reads out now=> apple
> > expected message=> apple is selected or selected apple.
> > 2).In mobile mode, on tab we can see 2 focus. it has to be 1 at a time.
> > 3).On tab we can see focus with blue border on message. Is it expected ?
> > 
> 
> > Now its reading properly.
> 
> > In Mobile mode Tab key won't present so we are discarding tab key access for mobile after. discussing with Sanjay , Puru and Sushma.

Hi Manu,

  Issues 2 and 3 are still the same . Tab focus is still present.

**Comment by manu-kr** [comment]
> > > 1)It reads out the name but does not mention that item is selected on select.
> > > when I select apple
> > > onselect reads out now=> apple
> > > expected message=> apple is selected or selected apple.
> > > 2).In mobile mode, on tab we can see 2 focus. it has to be 1 at a time.
> > > 3).On tab we can see focus with blue border on message. Is it expected ?
> > > 
> > 
> > 
> > > Now its reading properly.
> > 
> > 
> > > In Mobile mode Tab key won't present so we are discarding tab key access for mobile after. discussing with Sanjay , Puru and Sushma.
> 
> Hi Manu,
> 
> Issues 2 and 3 are still the same . Tab focus is still present.
> 

> FIxed now.

**Review by sKudum — APPROVED** [review]

**Comment by puru-hk** [comment]
@manu-kr Please find the below tested results by the accessibility team 

Issue: 1
The issue is partially fixed, the label is provided and the label is not read by the screen reader 
image

The associated label name should be read.

Issue 2: Issue fixed


Issue 3:
The focus is provided for the read only element. The tab stop should be removed for the disabled UI element 
The 'Read Only' information is not read.
image

Issue 4:
The dialog name is not read and closable information is not read.
No keyboard support provided for the 'Close' button.
image




Issue 5:
There is keyboard support provided for the list using the arrow keys, As focus goes to the last list item the slider doesn't don't go down the list selected is not visible.



Issue 6:
The pop over doesn't close using the 'ESC' key through out the application.
The pop over should be closed using the 'ESC' key.

Issue 7:
The Horizontal scroll bar is available when screen is resized to 200% zoom.
Horizontal scroll bar should be avoided
image

**Comment by Betrozov** [comment]
@manu-kr 
please squash all your commits in this PR and do one valid commit

**Comment by sKudum** [comment]
> > > > 1)It reads out the name but does not mention that item is selected on select.
> > > > when I select apple
> > > > onselect reads out now=> apple
> > > > expected message=> apple is selected or selected apple.
> > > > 2).In mobile mode, on tab we can see 2 focus. it has to be 1 at a time.
> > > > 3).On tab we can see focus with blue border on message. Is it expected ?
> > > > 
> > > 
> > > 
> > > > Now its reading properly.
> > > 
> > > 
> > > > In Mobile mode Tab key won't present so we are discarding tab key access for mobile after. discussing with Sanjay , Puru and Sushma.
> > 
> > 
> > Hi Manu,
> > Issues 2 and 3 are still the same . Tab focus is still present.
> > 
> 
> > FIxed now.

Hi Manu,

 After selecting a item and clicking on screen instead of dropdown button it is reopening the dialog until we click on close/cancel button.


https://user-images.githubusercontent.com/53487468/122914892-c46dd280-d378-11eb-824c-de6291369076.mp4

File: libs/core/src/lib/select/select-key-manager.service.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import { Injectable } from '@angular/core';
 import { takeUntil } from 'rxjs/operators';
 import { ActiveDescendantKeyManager } from '@angular/cdk/a11y';
-import { KeyUtil } from '@fundamental-ngx/core/utils';
-import { UP_ARROW, DOWN_ARROW, ENTER, SPACE, HOME, END, hasModifierKey } from '@angular/cdk/keycodes';
+import { KeyUtil } from '../utils/functions';
```
Comment (aberikashvili) [inline libs/core/src/lib/select/select-key-manager.service.ts:4]: Why `@fundamental-ngx/core/utils` is replaced by `../utils/functions`?

File: libs/core/src/lib/select/select.component.ts [hunk 2]
```diff
@@ -600,15 +602,20 @@ export class SelectComponent implements
...
+        const optionHeight = offsetHeight === 0  ? itemHeight : offsetHeight;
         const optionOffset = optionIndex * optionHeight;
+        panelHeight = SELECT_PANEL_MAX_HEIGHT;
         if (optionOffset < currentScrollPosition) {
             return optionOffset;
         }
-        if (optionOffset + optionHeight > currentScrollPosition + panelHeight) {
+
+       if (optionOffset + optionHeight > currentScrollPosition + panelHeight) {
+            console.log('sds', optionOffset - panelHeight + optionHeight);
```
Comment (aberikashvili) [inline libs/core/src/lib/select/select.component.ts:618]: Please remove console.log

File: libs/core/src/lib/select/select.component.ts [hunk 3]
```diff
@@ -600,15 +602,20 @@ export class SelectComponent implements
...
         optionIndex: number,
-        optionHeight: number,
+        itemHeight: number,
         currentScrollPosition: number,
         panelHeight: number
     ): number {
+        const offsetHeight = this._options.get(optionIndex)._getHtmlElement().offsetHeight ;
+        const optionHeight = offsetHeight === 0  ? itemHeight : offsetHeight;
         const optionOffset = optionIndex * optionHeight;
+        panelHeight = SELECT_PANEL_MAX_HEIGHT;
```
Comment (aberikashvili) [inline libs/core/src/lib/select/select.component.ts:612]: Why it's reassigned here?

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
Missing this bit from the original issue:

`The pop over doesn't close using the 'ESC' key through out the application.
The pop over should be closed using the 'ESC' key.`

Also I noticed that the screenreader incorrectly says the list has one item when you click the add-on button to open the options popover, but it works fine if you click the text to open it.

**Comment by droshev** [comment]
closing it as there was not much activities done. Please, reopen it once everything is done.

**Comment by manu-kr** [comment]
Opening this PR to do accessibility test in JAWS in VDI desktop

**Comment by droshev** [comment]
@manu-kr this PR is not correct. It contains commits that are not from you

File: libs/core/src/lib/select/select-key-manager.service.ts (hunk 1 above)
Comment (manu-kr) [reply libs/core/src/lib/select/select-key-manager.service.ts]: replaced.

File: libs/core/src/lib/select/select.component.ts (hunk 2 above)
Comment (manu-kr) [reply libs/core/src/lib/select/select.component.ts]: removed.

File: libs/core/src/lib/select/select.component.ts (hunk 3 above)
Comment (manu-kr) [reply libs/core/src/lib/select/select.component.ts]: removed.

**Comment by puru-hk** [comment]
issue fixed

**Review by puru-hk — APPROVED** [review]

**Comment by manu-kr** [comment]
> Missing this bit from the original issue:
> 
> `The pop over doesn't close using the 'ESC' key through out the application. The pop over should be closed using the 'ESC' key.`
> 
> Also I noticed that the screenreader incorrectly says the list has one item when you click the add-on button to open the options popover, but it works fine if you click the text to open it.



> Missing this bit from the original issue:
> 
> `The pop over doesn't close using the 'ESC' key through out the application. The pop over should be closed using the 'ESC' key.`
> 
> Also I noticed that the screenreader incorrectly says the list has one item when you click the add-on button to open the options popover, but it works fine if you click the text to open it.

@mikerodonnell89 Closing the popover in ESC key is fixed now.

**Review by Lokanathan-k — APPROVED** [review]
LGTM

**Review by platon-rov — APPROVED** [review]

**Review by mikerodonnell89 — CHANGES_REQUESTED** [review]
I am seeing some strange behavior when opening the mobile example. Note the focus outline surrounding the content.  When I close the select dialog with the escape key the focus remains, and then clicking a different select will re-open the mobile dialog.  This won't usually be an issue with touch screens as the user won't be using an escape key but it will be a problem if the user has a keyboard extension accessory, or if an application developer choses to implement everything in mobile/cozy mode
Screen Shot 2021-08-16 at 3 39 03 PM

**Review by Betrozov — APPROVED** [review]

**Comment by manu-kr** [comment]
> I am seeing some strange behavior when opening the mobile example. Note the focus outline surrounding the content. When I close the select dialog with the escape key the focus remains, and then clicking a different select will re-open the mobile dialog. This won't usually be an issue with touch screens as the user won't be using an escape key but it will be a problem if the user has a keyboard extension accessory, or if an application developer choses to implement everything in mobile/cozy mode
> Screen Shot 2021-08-16 at 3 39 03 PM

@mikerodonnell89 Issue is fixed now.

**Comment by droshev** [comment]
> > I am seeing some strange behavior when opening the mobile example. Note the focus outline surrounding the content. When I close the select dialog with the escape key the focus remains, and then clicking a different select will re-open the mobile dialog. This won't usually be an issue with touch screens as the user won't be using an escape key but it will be a problem if the user has a keyboard extension accessory, or if an application developer choses to implement everything in mobile/cozy mode
> > Screen Shot 2021-08-16 at 3 39 03 PM
> 
> @mikerodonnell89 Issue is fixed now.

@manu-kr looks like the QAs didn't catch this. I think the QA testing should restart.

**Review by mikerodonnell89 — APPROVED** [review]

**Comment by puru-hk** [comment]
Tested Looks fine

**Comment by manu-kr** [comment]
This PR is blocked until e2e test get fixed for date-picker and date-time-picker component.

**Comment by Seamoo13** [comment]
> This PR is blocked until e2e test get fixed for date-picker and date-time-picker component.

the fix is here:
https://github.com/SAP/fundamental-ngx/pull/6380

**Pull request merged**